### PR TITLE
add mixed-precision support for particle data

### PR DESCRIPTION
**Pull request by OsmanSeckinSimsek**
implemented the infrastructure for mixed-precision for particle data using variants. The file I/O for H5 and ascii also updated.

### CODE REVIEW

File: main/src/io/file_utils.hpp [hunk 1]
```diff
@@ -47,7 +48,9 @@ void writeAscii(size_t firstIndex, size_t lastIndex, const std::string& path, bo
             for (auto field : fields)
             {
                 [[maybe_unused]] std::initializer_list<int> list{(dumpFile << separators, 0)...};
-                dumpFile << field[i];
+                std::visit([&dumpFile, i](auto& arg) {
+                    dumpFile << arg[i];
+                }, field);
```
Comment (sekelle) [inline main/src/io/file_utils.hpp:53]: You should be able to pass `field[i]` directly here and avoid capture-by-value of `i`.

Comment (sekelle) [inline main/src/io/file_utils.hpp:53]: You should be able to pass `field[i]` directly here and avoid capture-by-value of `i`.

File: main/src/io/mpi_file_utils.hpp [hunk 2]
```diff
@@ -118,7 +131,10 @@ void writeH5Part(Dataset& d, size_t firstIndex, size_t lastIndex, const cstone::
     for (size_t fidx = 0; fidx < fieldPointers.size(); ++fidx)
     {
         const std::string& fieldName = Dataset::fieldNames[d.outputFields[fidx]];
-        writeH5PartField(h5_file, fieldName, fieldPointers[fidx] + firstIndex);
+        std::visit([&h5_file, &fieldName, fidx, firstIndex](auto& arg)
+        {
+            writeH5PartField(h5_file, fieldName, arg + firstIndex);
+        }, fieldPointers[fidx]);
```
Comment (sekelle) [inline main/src/io/mpi_file_utils.hpp:137]: `fidx` captured but not used?

Comment (sekelle) [inline main/src/io/mpi_file_utils.hpp:137]: `fidx` captured but not used?

File: sph/include/sph/data_util.hpp [hunk 3]
```diff
@@ -65,13 +65,17 @@ auto getOutputArrays(Dataset& dataset)
 {
     using T            = typename Dataset::RealType;
     auto fieldPointers = dataset.data();
+    using outputFieldType = std::variant<float*, double*, int*>;
```
Comment (sebkelle1) [inline sph/include/sph/data_util.hpp:68]: I would propose to name this simply `FieldType`

Comment (sebkelle1) [inline sph/include/sph/data_util.hpp:68]: I would propose to name this simply `FieldType`

File: main/src/io/file_utils.hpp [hunk 4]
```diff
@@ -27,9 +28,9 @@ void writeColumns(std::ostream& out, const Separator& sep, Columns&&... columns)
  * @param   fields         pointers to field array, each field is a column
  * @param   separators     arbitrary number of separators to insert between columns, eg '\t', std::setw(n), ...
  */
-template<class T, class... Separators>
+template<class... Separators>
 void writeAscii(size_t firstIndex, size_t lastIndex, const std::string& path, bool append,
-                const std::vector<T*>& fields, Separators&&... separators)
+                const std::vector<std::variant<float*, double*, int*>>& fields, Separators&&... separators)
```
Comment (sekelle) [inline main/src/io/file_utils.hpp]: You might be able to do `std::variant<Ts*...>` here with `class... Ts, class... Separators` to keep it generic.

Comment (sekelle) [inline main/src/io/file_utils.hpp]: You might be able to do `std::variant<Ts*...>` here with `class... Ts, class... Separators` to keep it generic.

File: sph/include/sph/particles_data.hpp [hunk 5]
```diff
@@ -33,6 +33,7 @@
 #include <cstdio>
 #include <iostream>
 #include <vector>
+#include
```
Comment (sekelle) [inline sph/include/sph/particles_data.hpp]: leftover?

Comment (sebkelle1) [inline sph/include/sph/particles_data.hpp:36]: leftover?

File: main/src/io/file_utils.hpp (hunk 1 above)
Comment (OsmanSeckinSimsek) [reply main/src/io/file_utils.hpp:53]: I tried to do that too but, as field is a variant, I am not able to dereference it directly there. Within the visit function when the type is known we can dereference. The compiler complains when I try to pass field[i].

File: sph/include/sph/particles_data.hpp (hunk 5 above)
Comment (OsmanSeckinSimsek) [reply sph/include/sph/particles_data.hpp]: it should have been <variant>, I'll fix that

File: sph/include/sph/data_util.hpp [hunk 6]
```diff
@@ -65,13 +65,17 @@ auto getOutputArrays(Dataset& dataset)
...
-
+    std::vector<outputFieldType> outputFields(dataset.outputFields.size());
+    int outIndex = 0;
+    for (int i : dataset.outputFields)
+    {
+        std::visit([&outputFields, outIndex](auto& arg) {
+            outputFields[outIndex] = arg->data();
+        }, fieldPointers[i]);
+        outIndex++;
+    }
```
Comment (sekelle) [inline sph/include/sph/data_util.hpp:79]:
```suggestion
std::vector<outputFieldType> outputFields;
outputFields.reserve(dataset.outputFields.size());
for (int i : dataset.outputFields)
{
    std::visit([&outputFields](auto& arg) {
        outputFields.push_back(arg->data());
    }, fieldPointers[i]);
}
    
```

Comment (sebkelle1) [inline sph/include/sph/data_util.hpp:78]:
```suggestion
std::vector<outputFieldType> outputFields;
outputFields.reserve(dataset.outputFields.size());

for (int i : dataset.outputFields)
{
    std::visit([&outputFields, outIndex](auto& arg) {
        outputFields.push_back(arg->data());
    }, fieldPointers[i]);
}
    
```

**Review by sekelle — CHANGES_REQUESTED** [review]
I'll test HDF5 and ASCII output, the CI does not have a very thorough coverage there.

**Comment by sekelle** [comment]
cscs-ci run

File: main/src/io/file_utils.hpp (hunk 4 above)
Comment (sekelle) [reply main/src/io/file_utils.hpp]: pointer types should be `const`

File: main/src/io/file_utils.hpp (hunk 1 above)
Comment (sekelle) [reply main/src/io/file_utils.hpp:53]: ah yes, correct.

File: sph/include/sph/data_util.hpp (hunk 3 above)
Comment (sekelle) [inline sph/include/sph/data_util.hpp]:
```suggestion
    using outputFieldType = std::variant<const float*, const double*, const int*>;
```

File: main/src/io/file_utils.hpp (hunk 4 above)
Comment (OsmanSeckinSimsek) [reply main/src/io/file_utils.hpp]: following our discussion on slack, I will update the code according to this

**Comment by sekelle** [comment]
cscs-ci run

**Comment by sekelle** [comment]
cscs-ci run

**Comment by sekelle** [comment]
cscs-ci run

**Review by sekelle — APPROVED** [review]

**Pull request merged**